iotbx/cif/builders.py

```diff
@@ -333,263 +337,28 @@ class miller_array_builder(crystal_symmetry_builder):
-                  if substr == '': continue
-                  if string.endswith(substr):
-                    string = string[:-len(substr)]
-                return string
-              # determine observation type
-              stripped_key = rstrip_substrings(
-                key, [key_suffix, '_au', '_meas', '_calc', '_plus', '_minus'])
-              if (stripped_key.endswith('F_squared') or
-                  stripped_key.endswith('intensity') or
-                  stripped_key.endswith('.I') or
-                  stripped_key.endswith('_I')) and (
-                    array.is_real_array() or array.is_integer_array()):
-                array.set_observation_type_xray_intensity()
-              elif (stripped_key.endswith('F') and (
-                array.is_real_array() or array.is_integer_array())):
-                array.set_observation_type_xray_amplitude()
-              if (array.is_xray_amplitude_array() or
-                  array.is_xray_amplitude_array()):
-                # e.g. merge_equivalents treats integer arrays differently, so must
-                # convert integer observation arrays here to be safe
-                if isinstance(array.data(), flex.int):
-                  array = array.customized_copy(data=array.data().as_double())
-              array.set_info(base_array_info.customized_copy(labels=labels))
-              if (array.is_xray_amplitude_array() or
-                  array.is_xray_amplitude_array()):
-                info = array.info()
-                array.set_info(info.customized_copy(wavelength=wavelength))
-              self._arrays.setdefault(key, array)
-    for key, array in six.iteritems(self._arrays.copy()):
-      if (   key.endswith('_minus') or '_minus_' in key
-          or key.endswith('_plus') or '_plus_' in key):
-        if '_minus' in key:
-          minus_key = key
-          plus_key = key.replace('_minus', '_plus')
-        elif '_plus' in key:
-          plus_key = key
-          minus_key = key.replace('_plus', '_minus')
-        if plus_key in self._arrays and minus_key in self._arrays:
-          plus_array = self._arrays.pop(plus_key)
-          minus_array = self._arrays.pop(minus_key)
-          minus_array = minus_array.customized_copy(
-            indices=-minus_array.indices()).set_info(minus_array.info())
-          array = plus_array.concatenate(
-            minus_array, assert_is_similar_symmetry=False)
-          array = array.customized_copy(anomalous_flag=True)
-          array.set_info(minus_array.info().customized_copy(
-            labels=list(
-              OrderedSet(plus_array.info().labels+minus_array.info().labels))))
-          array.set_observation_type(plus_array.observation_type())
-          self._arrays.setdefault(key, array)
-
-    if len(self._arrays) == 0:
-      raise CifBuilderError("No reflection data present in cif block")
-
+  def guess_observationtype(self, labl):
+    for okey in list(self.observation_types.keys()):
```

Comment:
These keys probably do not need to be in a list since the dictionary is not modified in the loop.